### PR TITLE
Rabin karp

### PR DESCRIPTION
**Pull request by archi14**
#789

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Sample Input / Output.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

#### Changes proposed in this pull request:
- Added Rabin karp code

#### Languages Used:
-Java 

#### Files Added:
- Rabin code java file.

### CODE REVIEW

File: Rabin_karp/Rabin_karp.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
+public class Main  
+{ 
+    // d is the number of characters in the input alphabet 
+    public final static int d = 256; 
+      
+      static void search(String pat, String txt, int q)
```
Comment (MastersAbh) [inline Rabin_karp/Rabin_karp.java]: There are 5 spaces before static.
Please fix indentation, use 4 spaces.

File: Rabin_karp/Rabin_karp.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    { 
+        int M = pat.length(); 
+        int N = txt.length(); 
+        int i, j; 
+        int p = 0; // hash value for pattern 
+        int t = 0; // hash value for txt 
+        int h = 1; 
+      
+        // The value of h would be "pow(d, M-1)%q" 
+        for (i = 0; i < M-1; i++)
```
Comment (MastersAbh) [inline Rabin_karp/Rabin_karp.java]: spaces around -

File: Rabin_karp/Rabin_karp.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        int M = pat.length(); 
+        int N = txt.length(); 
+        int i, j; 
+        int p = 0; // hash value for pattern 
+        int t = 0; // hash value for txt 
+        int h = 1; 
+      
+        // The value of h would be "pow(d, M-1)%q" 
+        for (i = 0; i < M-1; i++) 
+            h = (h*d)%q;
```
Comment (MastersAbh) [inline Rabin_karp/Rabin_karp.java]: space around operators

File: Rabin_karp/Rabin_karp.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+      
+        // The value of h would be "pow(d, M-1)%q" 
+        for (i = 0; i < M-1; i++) 
+            h = (h*d)%q; 
+      
+        // Calculate the hash value of pattern and first 
+        for (i = 0; i < M; i++) 
+        { 
+            p = (d*p + pat.charAt(i))%q; 
+            t = (d*t + txt.charAt(i))%q;
```
Comment (MastersAbh) [inline Rabin_karp/Rabin_karp.java]: space around * and % in line 22,23

File: Rabin_karp/Rabin_karp.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+      
+            // Check the hash values of current window of text 
+            // and pattern. If the hash values match then only 
+            // check for characters on by one 
+            if ( p == t ) 
+            { 
+                /* Check for characters one by one */
+                for (j = 0; j < M; j++) 
+                { 
+                    if (txt.charAt(i+j) != pat.charAt(j))
```
Comment (MastersAbh) [inline Rabin_karp/Rabin_karp.java]: space around operators

File: Rabin_karp/Rabin_karp.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+                } 
+      
+                // if p == t and pat[0...M-1] = txt[i, i+1, ...i+M-1] 
+                if (j == M) 
+                    System.out.println("Pattern found at index " + i); 
+            } 
+      
+            // Calculate hash value for next window of text: Remove 
+            // leading digit, add trailing digit 
+            if ( i < N-M )
```
Comment (MastersAbh) [inline Rabin_karp/Rabin_karp.java]: space around operators

File: Rabin_karp/Rabin_karp.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+                // if p == t and pat[0...M-1] = txt[i, i+1, ...i+M-1] 
+                if (j == M) 
+                    System.out.println("Pattern found at index " + i); 
+            } 
+      
+            // Calculate hash value for next window of text: Remove 
+            // leading digit, add trailing digit 
+            if ( i < N-M ) 
+            { 
+                t = (d*(t - txt.charAt(i)*h) + txt.charAt(i+M))%q;
```
Comment (MastersAbh) [inline Rabin_karp/Rabin_karp.java]: space around operators

File: Rabin_karp/Rabin_karp.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+            // Calculate hash value for next window of text: Remove 
+            // leading digit, add trailing digit 
+            if ( i < N-M ) 
+            { 
+                t = (d*(t - txt.charAt(i)*h) + txt.charAt(i+M))%q; 
+      
+                // We might get negative value of t, converting it 
+                // to positive 
+                if (t < 0) 
+                t = (t + q);
```
Comment (MastersAbh) [inline Rabin_karp/Rabin_karp.java]: Indentation

**Comment by archi14** [comment]
@MastersAbh made the changes.

**Comment by archi14** [comment]
@MastersAbh Done.

File: Rabin_karp/Rabin_karp.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
+/*The Rabin–Karp algorithm is a string-searching algorithm that uses hashing to find any one of a set of pattern strings in a text.*/
```
Comment (MastersAbh) [inline Rabin_karp/Rabin_karp.java]: Please split it into multiple lines and we're done.

**Comment by archi14** [comment]
@MastersAbh Done.

File: Rabin_karp/Rabin_karp.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
+/*The Rabin–Karp algorithm is a string-searching algorithm that uses
+hashing to find any one of a set of pattern strings in a text.*/
+public class Main
```
Comment (MastersAbh) [inline Rabin_karp/Rabin_karp.java]: one blank line before class declaration.

**Pull request merged**